torcheeg/models/transformer/retnet.py

```diff
@@ -0,0 +1,359 @@
+        r_is = []
+        for j in range(self.heads):
+            y, r_i = self.retentions[j].forward_chunkwise(
+                x_i[:, :, :], r_i_1s[j], i
+                )
+            Y.append(y)
+            r_is.append(r_i)
+        
+        
+        Y = torch.cat(Y, dim=2)
+        Y_shape = Y.shape
+        Y = self.group_norm(Y.reshape(-1, self.v_dim)).reshape(Y_shape)
+
+        return (self.swish(x_i @ self.W_G) * Y) @ self.W_O, r_is
+
+
+class RetNet(nn.Module):
+    def __init__(self, layers, hidden_dim, ffn_size, heads, double_v_dim=False):
+        super(RetNet, self).__init__()
+        self.layers = layers
+        self.hidden_dim = hidden_dim
+        self.ffn_size = ffn_size
+        self.heads = heads
+        self.v_dim = hidden_dim * 2 if double_v_dim else hidden_dim
+
+        self.retentions = nn.ModuleList([
+            MultiScaleRetention(hidden_dim, heads, double_v_dim)
+            for _ in range(layers)
+        ])
+        self.ffns = nn.ModuleList([
+            nn.Sequential(
+                nn.Linear(hidden_dim, ffn_size),
+                nn.GELU(),
+                nn.Linear(ffn_size, hidden_dim)
+            )
+            for _ in range(layers)
+        ])
+        self.layer_norms_1 = nn.ModuleList([
+            nn.LayerNorm(hidden_dim)
+            for _ in range(layers)
+        ])
+        self.layer_norms_2 = nn.ModuleList([
+            nn.LayerNorm(hidden_dim)
+            for _ in range(layers)
+        ])
+    
+    def forward(self, X):
+        """
+        X: (batch_size, sequence_length, hidden_size)
+        """
+        for i in range(self.layers):
+            Y = self.retentions[i](self.layer_norms_1[i](X)) + X
+           
+            X = self.ffns[i](self.layer_norms_2[i](Y)) + Y
+
+        return X
+
+    def forward_recurrent(self, x_n, s_n_1s, n):
+        """
+        X: (batch_size, sequence_length, hidden_size)
+        s_n_1s: list of lists of tensors of shape (batch_size, hidden_size // heads, hidden_size // heads)
+
+        """
+        s_ns = []
+        for i in range(self.layers):
+            # list index out of range
+            o_n, s_n = self.retentions[i].forward_recurrent(self.layer_norms_1[i](x_n), s_n_1s[i], n)
+            y_n = o_n + x_n
+            s_ns.append(s_n)
+            x_n = self.ffns[i](self.layer_norms_2[i](y_n)) + y_n
+        
+        return x_n, s_ns
+    
+    def forward_chunkwise(self, x_i, r_i_1s, i):
+        """
+        X: (batch_size, sequence_length, hidden_size)
+        r_i_1s: list of lists of tensors of shape (batch_size, hidden_size // heads, hidden_size // heads)
+
+        """
+        r_is = []
+        for j in range(self.layers):
+            o_i, r_i = self.retentions[j].forward_chunkwise(self.layer_norms_1[j](x_i), r_i_1s[j], i)
+            y_i = o_i + x_i
+            r_is.append(r_i)
+            x_i = self.ffns[j](self.layer_norms_2[j](y_i)) + y_i
+        
+        return x_i, r_is
+    if __name__ == "__main__":
```

Comment:
This part can be deleted. Please add test cases to https://github.com/torcheeg/torcheeg/blob/main/test/models/test_transformer.py.